x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/TimeSeriesSortedSourceOperatorFactory.java

```diff
@@ -0,0 +1,343 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.compute.lucene;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.SortedDocValues;
+import org.apache.lucene.index.SortedNumericDocValues;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.PriorityQueue;
+import org.elasticsearch.compute.data.BlockFactory;
+import org.elasticsearch.compute.data.DocVector;
+import org.elasticsearch.compute.data.IntBlock;
+import org.elasticsearch.compute.data.IntVector;
+import org.elasticsearch.compute.data.LongVector;
+import org.elasticsearch.compute.data.Page;
+import org.elasticsearch.compute.operator.DriverContext;
+import org.elasticsearch.compute.operator.SourceOperator;
+import org.elasticsearch.core.Releasables;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.List;
+import java.util.function.Function;
+
+/**
+ * Creates a source operator that takes advantage of the natural sorting of segments in a tsdb index.
+ * <p>
+ * This source operator loads the _tsid and @timestamp fields, which is used for emitting documents in the correct order. These field values
+ * are included in the page as seperate blocks and downstream operators can make use of these loaded time series ids and timestamps.
+ * <p>
+ * The source operator includes all documents of a time serie with the same page. So the same time series never exists in multiple pages.
+ * Downstream operators can make use of this implementation detail.
+ * <p>
+ * This operator currently only supports shard level concurrency. A new concurrency mechanism should be introduced at the time serie level
+ * in order to read tsdb indices in parallel.
```

Comment:
> Great point - let's make sure we don't lose track of this.

It is part of the #105397

> What happens in case of doc_partitioning is used?

It is ignored.